iceoryx2/src/port/waitset.rs

```diff
@@ -0,0 +1,718 @@
+//! cyclic wakeup timeouts (Ticks) can be attached.
+//!
+//! # Example
+//!
+//! ## Notification
+//!
+//! ```no_run
+//! use iceoryx2::prelude::*;
+//! # use core::time::Duration;
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let node = NodeBuilder::new().create::<ipc::Service>()?;
+//! # let event = node.service_builder(&"MyEventName_1".try_into()?)
+//! #     .event()
+//! #     .open_or_create()?;
+//!
+//! let mut listener = event.listener_builder().create()?;
+//!
+//! let waitset = WaitSetBuilder::new().create::<ipc::Service>()?;
+//! let guard = waitset.attach_notification(&listener)?;
+//!
+//! let event_handler = |attachment_id: AttachmentId<ipc::Service>| {
+//!     if attachment_id.event_from(&guard) {
+//!         while let Ok(Some(event_id)) = listener.try_wait_one() {
+//!             println!("received notification {:?}", event_id);
+//!         }
+//!     }
+//! };
+//!
+//! while waitset.run(event_handler).is_ok() {}
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ## Deadline
+//!
+//! ```no_run
+//! use iceoryx2::prelude::*;
+//! # use core::time::Duration;
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let node = NodeBuilder::new().create::<ipc::Service>()?;
+//! # let event = node.service_builder(&"MyEventName_1".try_into()?)
+//! #     .event()
+//! #     .open_or_create()?;
+//!
+//! let mut listener = event.listener_builder().create()?;
+//!
+//! let listener_deadline = Duration::from_secs(1);
+//! let waitset = WaitSetBuilder::new().create::<ipc::Service>()?;
+//! let guard = waitset.attach_deadline(&listener, listener_deadline)?;
+//!
+//! let event_handler = |attachment_id: AttachmentId<ipc::Service>| {
+//!     if attachment_id.event_from(&guard) {
+//!         while let Ok(Some(event_id)) = listener.try_wait_one() {
+//!             println!("received notification {:?}", event_id);
+//!         }
+//!     } else if attachment_id.deadline_from(&guard) {
+//!         println!("Oh no, we hit the deadline without receiving any kind of event");
+//!     }
+//! };
+//!
+//! while waitset.run(event_handler).is_ok() {}
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ## Tick
+//!
+//! ```no_run
+//! use iceoryx2::prelude::*;
+//! # use core::time::Duration;
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let node = NodeBuilder::new().create::<ipc::Service>()?;
+//! # let pubsub = node.service_builder(&"MyServiceName".try_into()?)
+//! #     .publish_subscribe::<u64>()
+//! #     .open_or_create()?;
+//!
+//! let publisher_1 = pubsub.publisher_builder().create()?;
+//! let publisher_2 = pubsub.publisher_builder().create()?;
+//!
+//! let pub_1_period = Duration::from_millis(250);
+//! let pub_2_period = Duration::from_millis(718);
+//!
+//! let waitset = WaitSetBuilder::new().create::<ipc::Service>()?;
+//! let guard_1 = waitset.attach_tick(pub_1_period)?;
+//! let guard_2 = waitset.attach_tick(pub_2_period)?;
```

Comment:
Food for thought. Would one attach an interval and get a tick? Similar to attaching a deadline and getting a timeout.

Reply:
Yes this is not a "tick"...

Reply:
So would you all be happy with `attach_interval` and whenever it comes it is a tick? or a wakeup?

Reply:
Sounds more natural to me. The confusion is that when I see tick, I think of the smallest individual increment of a clock.